File: backend/app/schemas/statistics.py
"""
统计数据相关的Pydantic模式
"""
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


class ProjectStats(BaseModel):
    """项目统计信息"""
    project_id: int = Field(..., description="项目ID")
    project_name: str = Field(..., description="项目名称")
    total_mrs: int = Field(..., description="总MR数量")
    merged_mrs: int = Field(..., description="已合并MR数量")
    open_mrs: int = Field(..., description="打开状态MR数量")
    total_additions: int = Field(..., description="总新增行数")
    total_deletions: int = Field(..., description="总删除行数")
    net_changes: int = Field(..., description="净变化行数")
    avg_mr_size: float = Field(..., description="平均MR大小")
    review_coverage: float = Field(..., description="审查覆盖率")
    # 新增：代码增长和减少的详细指标
    code_growth_rate: float = Field(..., description="代码增长率(%)")
    code_reduction_rate: float = Field(..., description="代码减少率(%)")
    avg_additions_per_mr: float = Field(..., description="平均每个MR新增行数")
    avg_deletions_per_mr: float = Field(..., description="平均每个MR删除行数")


class UserStats(BaseModel):
    """用户统计信息"""
    author: str = Field(..., description="作者")
    total_mrs: int = Field(..., description="总MR数量")
    merged_mrs: int = Field(..., description="已合并MR数量")
    total_additions: int = Field(..., description="总新增行数")
    total_deletions: int = Field(..., description="总删除行数")
    net_changes: int = Field(..., description="净变化行数")
    avg_mr_size: float = Field(..., description="平均MR大小")
    avg_review_score: Optional[float] = Field(None, description="平均审查评分")
    total_commits: int = Field(..., description="总提交数")
    # 新增：用户代码变更详细指标
    code_growth_rate: float = Field(..., description="代码增长率(%)")
    code_reduction_rate: float = Field(..., description="代码减少率(%)")
    avg_additions_per_mr: float = Field(..., description="平均每个MR新增行数")
    avg_deletions_per_mr: float = Field(..., description="平均每个MR删除行数")
    productivity_score: float = Field(..., description="生产力评分(基于代码变更量和质量)")
    # 新增：分层生产力评分
    contribution_score: float = Field(..., description="贡献层评分(40%)")
    code_efficiency_score: float = Field(..., description="代码效率层评分(30%)")
    quality_score: float = Field(..., description="质量层评分(20%)")
    participation_score: float = Field(..., description="参与活跃层评分(10%)")


class CodeQualityTrend(BaseModel):
    """代码质量趋势"""
    date: str = Field(..., description="日期")
    avg_score: Optional[float] = Field(None, description="平均评分")
    review_count: int = Field(..., description="审查数量")
    pass_rate: float = Field(..., description="通过率")


class EfficiencyMetrics(BaseModel):
    """效率指标"""
    avg_mr_duration: float = Field(..., description="平均MR处理时间(小时)")
    avg_review_time: float = Field(..., description="平均审查时间(小时)")
    mrs_per_day: float = Field(..., description="每日MR数量")
    commits_per_day: float = Field(..., description="每日提交数")


class TechnicalDebt(BaseModel):
    """技术债务"""
    long_pending_mrs: int = Field(..., description="长期待处理MR数量")
    re_reviewed_mrs: int = Field(..., description="重复审查MR数量")
    high_risk_patterns: int = Field(..., description="高风险代码模式数量")
    oldest_pending_days: Optional[int] = Field(None, description="最老待处理MR天数")


class TokenUsageTrend(BaseModel):
    """Token使用趋势"""
    date: str = Field(..., description="日期")
    total_tokens: int = Field(..., description="总token数")
    direct_tokens: int = Field(..., description="直接token数")
    cache_tokens: int = Field(..., description="缓存token数")
    prompt_tokens: int = Field(..., description="输入token数")
    completion_tokens: int = Field(..., description="输出token数")
    cost: float = Field(..., description="成本")


class TokenUsageByModel(BaseModel):
    """按模型统计Token使用"""
    model_name: str = Field(..., description="模型名称")
    total_tokens: int = Field(..., description="总token数")
    direct_tokens: int = Field(..., description="直接token数")
    cache_tokens: int = Field(..., description="缓存token数")
    cost: float = Field(..., description="成本")
    usage_count: int = Field(..., description="使用次数")


class TokenUsageByProject(BaseModel):
    """按项目统计Token使用"""
    project_name: str = Field(..., description="项目名称")
    total_tokens: int = Field(..., description="总token数")
    direct_tokens: int = Field(..., description="直接token数")
    cache_tokens: int = Field(..., description="缓存token数")
    cost: float = Field(..., description="成本")
    review_count: int = Field(..., description="审查次数")


# 注意：StatisticsResponse 和 StatisticsRequest 已废弃
# 现在使用分接口，每个接口有自己的请求和响应格式

File: frontend/src/api/statistics.ts
import api from './index'

// 类型定义
export interface TimeRangeFilter {
  start_date?: string
  end_date?: string
  period?: 'today' | 'this_week' | 'this_month' | 'week' | 'month' | 'quarter' | 'year' | 'custom'
}

export interface ProjectStats {
  project_id: number
  project_name: string
  total_mrs: number
  merged_mrs: number
  open_mrs: number
  total_additions: number
  total_deletions: number
  net_changes: number
  avg_mr_size: number
  review_coverage: number
  code_growth_rate: number
  code_reduction_rate: number
  avg_additions_per_mr: number
  avg_deletions_per_mr: number
}

export interface UserStats {
  author: string
  total_mrs: number
  merged_mrs: number
  total_additions: number
  total_deletions: number
  net_changes: number
  avg_mr_size: number
  avg_review_score?: number
  total_commits: number
  code_growth_rate: number
  code_reduction_rate: number
  avg_additions_per_mr: number
  avg_deletions_per_mr: number
  productivity_score: number
}

export interface CodeQualityTrend {
  date: string
  avg_score?: number
  review_count: number
  pass_rate: number
}

export interface EfficiencyMetrics {
  avg_mr_duration: number
  avg_review_time: number
  mrs_per_day: number
  commits_per_day: number
}

export interface TechnicalDebt {
  long_pending_mrs: number
  re_reviewed_mrs: number
  high_risk_patterns: number
  oldest_pending_days?: number
}

export interface TokenUsageTrend {
  date: string
  total_tokens: number
  direct_tokens: number
  cache_tokens: number
  prompt_tokens: number
  completion_tokens: number
  cost: number
}

export interface TokenUsageByModel {
  model_name: string
  total_tokens: number
  direct_tokens: number
  cache_tokens: number
  cost: number
  usage_count: number
}

export interface TokenUsageByProject {
  project_name: string
  total_tokens: number
  direct_tokens: number
  cache_tokens: number
  cost: number
  review_count: number
}

export interface StatisticsResponse {
  time_range: TimeRangeFilter
  project_stats: ProjectStats[]
  user_stats: UserStats[]
  quality_trends: CodeQualityTrend[]
  efficiency_metrics: EfficiencyMetrics
  technical_debt: TechnicalDebt
  token_usage_trends: TokenUsageTrend[]
  token_usage_by_model: TokenUsageByModel[]
  token_usage_by_project: TokenUsageByProject[]
  summary: {
    total_projects: number
    total_mrs: number
    total_reviewed: number
    review_coverage: number
    avg_review_score: number
    total_authors: number
    period_days: number
    total_tokens: number
    total_direct_tokens: number
    total_cache_tokens: number
    total_cost: number
  }
}

// 基础统计API
export const basicStatsApi = {
  async getSummary() {
    const response = await api({
      method: 'GET',
      url: '/basic-stats/summary'
    })
    return response
  }
}

// Token统计API
export const tokenStatsApi = {
  async getSummary(days: number = 30) {
    const response = await api({
      method: 'GET',
      url: '/token-stats/summary',
      params: { days }
    })
    return response
  },

  async getTrends(days: number = 30) {
    const response = await api({
      method: 'GET',
      url: '/token-stats/trends',
      params: { days }
    })
    return response
  },

  async getByModel(days: number = 30) {
    const response = await api({
      method: 'GET',
      url: '/token-stats/by-model',
      params: { days }
    })
    return response
  },

  async getByProject(days: number = 30) {
    const response = await api({
      method: 'GET',
      url: '/token-stats/by-project',
      params: { days }
    })
    return response
  }
}

// 项目统计API
export const projectStatsApi = {
  async getStatistics(params: {
    period?: string
    project_ids?: string
    authors?: string
    time_criteria?: string
  } = {}) {
    const response = await api({
      method: 'GET',
      url: '/project-stats/',
      params
    })
    return response
  }
}

// 用户统计API
export const userStatsApi = {
  async getStatistics(params: {
    period?: string
    project_ids?: string
    authors?: string
    time_criteria?: string
  } = {}) {
    const response = await api({
      method: 'GET',
      url: '/user-stats/',
      params
    })
    return response
  }
}

// 质量统计API
export const qualityStatsApi = {
  async getTrends(params: {
    period?: string
    project_ids?: string
    authors?: string
    time_criteria?: string
  } = {}) {
    const response = await api({
      method: 'GET',
      url: '/quality-stats/trends',
      params
    })
    return response
  },

  async getTechnicalDebt(params: {
    period?: string
    project_ids?: string
    authors?: string
    time_criteria?: string
  } = {}) {
    const response = await api({
      method: 'GET',
      url: '/quality-stats/technical-debt',
      params
    })
    return response
  }
}

// 效率统计API
export const efficiencyStatsApi = {
  async getMetrics(params: {
    period?: string
    project_ids?: string
    authors?: string
    time_criteria?: string
  } = {}) {
    const response = await api({
      method: 'GET',
      url: '/efficiency-stats/',
      params
    })
    return response
  }
}

// 组合统计API - 一次性获取所有数据
export const combinedStatsApi = {
  async getAllStatistics(params: {
    period?: string
    project_ids?: string
    authors?: string
    time_criteria?: string
    days?: number
  } = {}) {
    const { period = 'this_month', project_ids, authors, time_criteria = 'activity', days = 30 } = params
    
    const commonParams = { period, project_ids, authors, time_criteria }
    
    // 并行请求所有接口
    const [
      basicSummary,
      tokenSummary,
      tokenTrends,
      tokenByModel,
      tokenByProject,
      projectStats,
      userStats,
      qualityTrends,
      technicalDebt,
      efficiencyMetrics
    ] = await Promise.all([
      basicStatsApi.getSummary(),
      tokenStatsApi.getSummary(days),
      tokenStatsApi.getTrends(days),
      tokenStatsApi.getByModel(days),
      tokenStatsApi.getByProject(days),
      projectStatsApi.getStatistics(commonParams),
      userStatsApi.getStatistics(commonParams),
      qualityStatsApi.getTrends(commonParams),
      qualityStatsApi.getTechnicalDebt(commonParams),
      efficiencyStatsApi.getMetrics(commonParams)
    ])

    // 组合返回数据，保持与原接口兼容的格式
    return {
      time_range: { period, start_date: null, end_date: null },
      project_stats: projectStats,
      user_stats: userStats,
      quality_trends: qualityTrends,
      efficiency_metrics: efficiencyMetrics,
      technical_debt: technicalDebt,
      token_usage_trends: tokenTrends,
      token_usage_by_model: tokenByModel,
      token_usage_by_project: tokenByProject,
      summary: {
        total_projects: basicSummary.total_projects,
        total_mrs: basicSummary.total_merge_requests,
        total_reviewed: basicSummary.reviewed_merge_requests,
        review_coverage: basicSummary.review_coverage,
        avg_review_score: (() => {
          // 从用户统计中计算平均得分
          const usersWithScore = userStats.filter((user: UserStats) => user.avg_review_score && user.avg_review_score > 0)
          if (usersWithScore.length === 0) return 0
          const totalScore = usersWithScore.reduce((sum: number, user: UserStats) => sum + (user.avg_review_score || 0), 0)
          return Math.round(totalScore / usersWithScore.length * 100) / 100 // 保留两位小数
        })(),
        total_authors: userStats.length,
        period_days: days,
        // Token相关汇总
        total_tokens: tokenSummary.total_tokens,
        total_direct_tokens: tokenSummary.total_direct_tokens,
        total_cache_tokens: tokenSummary.total_cache_tokens,
        total_cost: tokenSummary.total_cost
      }
    }
  }
}

export default {
  basicStatsApi,
  tokenStatsApi,
  projectStatsApi,
  userStatsApi,
  qualityStatsApi,
  efficiencyStatsApi,
  combinedStatsApi
}
